File: algorithms/java/src/test/java/leetcode/LC165CompareVersionNumbers.java
package leetcode;

import org.junit.jupiter.api.Test;


import static org.junit.jupiter.api.Assertions.assertEquals;

/**
 * https://leetcode.com/problems/compare-version-numbers/
 * #medium
 */
public class LC165CompareVersionNumbers {
    public class Solution {
        public int compareVersion(String version1, String version2) {
            String[] parts1 = version1.split("\\.");
            String[] parts2 = version2.split("\\.");
            for (int i = 0; i < Math.max(parts1.length, parts2.length); i++) {
                int level1 = 0;
                if (i < parts1.length) {
                    level1 = Integer.parseInt(parts1[i]);
                }
                int level2 = 0;
                if (i < parts2.length) {
                    level2 = Integer.parseInt(parts2[i]);
                }
                if (level1 < level2) {
                    return -1;
                }
                if (level1 > level2) {
                    return 1;
                }
            }
            return 0;
        }
    }

    @Test
    public void test_1__1() throws Exception {
        assertEquals(0, new Solution().compareVersion("1", "1"));
    }

    @Test
    public void test_1__1_0() throws Exception {
        assertEquals(0, new Solution().compareVersion("1", "1.0"));
    }

    @Test
    public void test_2__1() throws Exception {
        assertEquals(1, new Solution().compareVersion("2", "1"));
    }

    @Test
    public void test_1__13_1() throws Exception {
        assertEquals(-1, new Solution().compareVersion("1", "13.1"));
    }
}

File: algorithms/code/leetcode/lc165_compare_version_numbers/lc165_compare_version_numbers.py
#!/usr/bin/env python3
# https://leetcode.com/problems/compare-version-numbers/

import unittest


class Solution:
    def compareVersion(self, version1: str, version2: str) -> int:
        parts1 = version1.split('.')
        parts2 = version2.split('.')
        for i in range(max(len(parts1), len(parts2))):
            level1 = 0
            if i < len(parts1):
                level1 = int(parts1[i])
            level2 = 0
            if i < len(parts2):
                level2 = int(parts2[i])
            if level1 < level2:
                return -1
            if level1 > level2:
                return 1
        return 0


class TestCode(unittest.TestCase):
    def test_1__1(self):
        self.assertEqual(0, Solution().compareVersion('1', '1'))

    def test_1__1_0(self):
        self.assertEqual(0, Solution().compareVersion('1', '1.0'))

    def test_2__1(self):
        self.assertEqual(1, Solution().compareVersion('2', '1'))

    def test_1__13_1(self):
        self.assertEqual(-1, Solution().compareVersion('1', '13.1'))

    def test_1_0_1__1(self):
        self.assertEqual(1, Solution().compareVersion('1.0.1', '1'))
